src/lib.rs

```diff
@@ -234,6 +234,38 @@ impl GetAdjacencyMatrix for PyDAG {
     }
 }
 
+// Rust side only PyDAG methods
+impl PyDAG {
+    fn _add_edge(
+        &mut self,
+        p_index: NodeIndex,
+        c_index: NodeIndex,
+        edge: PyObject,
+    ) -> PyResult<usize> {
+        if self.check_cycle {
+            let should_check_for_cycle =
+                must_check_for_cycle(self, p_index, c_index);
```

Comment:
I remember looking through this and understanding it at some point, but it's confusing to see "check cycle" then "should check" then "must check". Maybe you could add a note about why these different check booleans are all necessary?

Reply:
Sure, I'll do that now. But to quickly summarize `check_cycle` is an instance attribute that says whether the cycle checking is run or not. It's basically a performance optimization because cycle checking gets increasingly expensive as the graph grows so we default it off (and for terra it's impossible for dagcircuit to introduce a cycle so this is safe) and make cycle checking opt-in. `must_check_for_cycle` and `should_check_for_cycle` should probably be renamed for clarity they're bad names. `must_check_for_cycle` is the function to check the relative position of parent and child in the graph: https://github.com/Qiskit/retworkx/blob/master/src/lib.rs#L661-L671 and says if the parent node has no parents or the child node has no children or there is an existing edge between them then it is safe to skip cycle checking because adding an edge couldn't introduce a cycle. `should_check_for_cycle` just is the output from that function in a variable. Bu I agree this is needlessly confusing

Reply:
Done in: https://github.com/Qiskit/retworkx/pull/82/commits/98d316eeaff8db546fbec33f80370279de193cd9